src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -186,8 +186,13 @@ class JvmtiTagHashmap : public CHeapObj<mtInternal> {
 
   // hash a given key (oop) with the specified size
   static unsigned int hash(oop key, int size) {
-    ZGC_ONLY(assert(ZAddressMetadataShift >= sizeof(unsigned int) * BitsPerByte, "cast removes the metadata bits");)
-
+#if INCLUDE_ZGC
+    if (UseZGC) {
+      const oop obj = Access<>::resolve(key);
+      const unsigned int hash = Universe::heap()->hash_oop(obj);
+      return hash % size;
+    }
+#endif
     // shift right to get better distribution (as these bits will be zero
```

Comment:
is it better to use #elif ... #endif to enclose line 196-203?

Reply:
if UseZGC is disabled, this function will become empty

Reply:
my code isolation methodology:

if the original patch is
```
- A
+ B
```
we will modify the snippet in this way
```
- A
+ #if INCLUDE_ZGC
+ if (UseZGC)
+   B
+ else
+ #endif
+   A
```

that make sure
* if ZGC compilation is enable
  * if UseZGC: run B
  * if !UseZGC: run A
* only compile A